[WP Individual JP Plugin] Remote overlay shown max count

thomashorta opened the pull request:
Part of #18175 

- Fix `RemoteConfigField` defaults loading to local database
- Create a remote field for getting the overlay max shown value

To test:
Logout, then run any logged in scenario from https://github.com/wordpress-mobile/WordPress-Android/pull/18169, such as opening the site picker having at least a problem site connected to your WordPress.com account, multiple times and **verify** the Individual Jetpack Plugin overlay modal is shown at most 3 times while logged in.

Also, verify that logging out and logging in again should clear the count so logging in and testing again will make the dialog be shown again 3 times.

The "3 times" value is currently being returned from the local default, we can also update the endpoint to return this remote config and do an integrated test (e.g.: set the endpoint to return 5 and **verify** the modal is shown 5 times).

## Regression Notes
1. Potential unintended areas of impact
Breakage/overwrite of other remote fields.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing.

3. What automated tests I added (or what prevented me from doing so)
Update existing tests.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Commit:
create max shown remote config

Commit:
remove default from RemoteConfigField class

It was not being used since the default actually comes from the
RemoteFieldDefaultGenerater annotation. So better to remove it to avoid
confusion.

Commit:
fix remote field repository to properly add defaults

Field defaults were ONLY being added when the local remote field DB table was
empty, but in cases of app update, that table will already have some values and
any new remote configs were not being added to it.

This fix makes sure to add missing fields if none was found in the currently
existing table.

Commit:
use remote config for max shown count

Commit:
fix remote config field type issue

Commit:
create last shown timestamp field in AppPrefs

Commit:
implement delay between overlays using AppPrefs

Commit:
insert remote config default if no value returned from server

this makes sure that defaults updated in code are saved to the database if no
value is returned to those keys from the server. It also makes sure we revert
to using local defaults if a remote config is ever deleted from the server.

Commit:
add unit tests for new scenarios

RenanLukas commented on WordPress/src/main/java/org/wordpress/android/util/config/BlazeRemoteFields.kt:
```diff
@@ -15,8 +15,7 @@ const val BLAZE_NON_DISMISSABLE_HASH_DEFAULT = "step-4"
     RemoteConfigField<String>(
         appConfig,
-        BLAZE_NON_DISMISSABLE_HASH_REMOTE_FIELD,
-        BLAZE_NON_DISMISSABLE_HASH_DEFAULT
```
❓ Just to confirm: should we remove the default parameters for `RemoteConfigField`? In this case but also [this](https://github.com/wordpress-mobile/WordPress-Android/pull/18189/files#diff-a74109012289f9f7a6c6c9dd3f48b21c5d446bae6291f60ba06ce02f0d82b8bbL34), [this](https://github.com/wordpress-mobile/WordPress-Android/pull/18189/files#diff-5b22ab944bd560974e9ed1c576b133be37681814e5737c20821aad3d38a983a4L14), [this](https://github.com/wordpress-mobile/WordPress-Android/pull/18189/files#diff-5b22ab944bd560974e9ed1c576b133be37681814e5737c20821aad3d38a983a4L25) and all the others.

thomashorta replied:
It was not being used at all in the class since the actual default is set by the `RemoteFieldDefaultGenerater` annotation. So it was rather confusing to pass the default twice.

RenanLukas commented on WordPress/src/main/java/org/wordpress/android/util/config/RemoteConfigField.kt:
```diff
@@ -1,13 +1,13 @@
 package org.wordpress.android.util.config
 
-open class RemoteConfigField<T>(val appConfig: AppConfig, val remoteField: String, val defaultValue: T) {
+open class RemoteConfigField<T>(val appConfig: AppConfig, val remoteField: String) {
```
❓ Do we still need the generic `T` after the changes to this class?

thomashorta replied:
We technically don't but I think it adds a good indicator of what the class expects, so I just pushed a change making sure it is used in `getValue`.

RenanLukas reviewed (commented):
Thanks for the changes, @thomashorta . It's working as expected 👍 

I've left a couple of questions.

ovitrif committed:
Merge pull request #18190 from wordpress-mobile/issue/18175-wp-jp-plugin-overlay-recurrence-delay

[WP Individual JP Plugin] Delay before overlay appearances

Commit:
change getValue to use the class generic type

Some changes were needed in the JetpackFeatureRemovalBrandingUtil since it was
expecting a `null` result sometimes, though that never actually happens, which
could even cause a crash. This fixes that issue and avoid crashes.

thomashorta commented:
Note: another PR was merged in this branch but it is not directly related to the max count.

RenanLukas reviewed (approved):
LGTM :shipit:
